app/src/org/commcare/views/widgets/CommCareAudioWidget.java

```diff
@@ -105,88 +114,180 @@ public IAnswerData getAnswer() {
-            recorder.setArguments(args);
         }
+        args.putString(APPEARANCE_ATTR_ARG_KEY, prompt.getAppearanceHint());
+        recorder.setArguments(args);
         recorder.show(((FragmentActivity)getContext()).getSupportFragmentManager(), "Recorder");
     }
 
     @Override
     public void setBinaryData(Object binaryuri) {
         super.setBinaryData(binaryuri);
-        if (recordedFileName != null) {
-            recordingNameText.setText(recordedFileName);
-        }
     }
 
     @Override
     public void onRecordingCompletion(String audioFile) {
-        setBinaryData(audioFile);
-        mPlayButton.setEnabled(true);
-        mPlayButton.setBackgroundResource(R.drawable.play);
-        recordingNameText.setTextColor(getResources().getColor(R.color.black));
-        recordingNameText.setText(Localization.get("recording.custom"));
+        if (new File(audioFile).exists()) {
+            setBinaryData(audioFile);
+            togglePlayButton(true);
+        } else {
+            clearAnswer();
+        }
     }
 
     @Override
     protected void playAudio() {
-        Uri filePath = Uri.parse(mInstanceFolder + mBinaryName);
-        player = MediaPlayer.create(getContext(), filePath);
-        player.setOnCompletionListener(mp -> resetAudioPlayer());
-        player.start();
         mPlayButton.setBackgroundResource(R.drawable.pause);
         mPlayButton.setOnClickListener(v -> pauseAudioPlayer());
+        startPlaybackTimer();
+        player.start();
+    }
+
+    private void startPlaybackTimer() {
+        playbackTimer = new Timer();
+        playbackTimer.schedule(new TimerTask() {
+            @Override
+            public void run() {
+                ((AppCompatActivity)getContext()).runOnUiThread(() -> {
+                    if (playbackTimer != null) {
+                        updatePlaybackInfo();
+                    }
+                });
+            }
+        }, 1 * 1000, 1 * 1000);
+    }
+
+    private void stopPlaybackTimer() {
+        if (playbackTimer != null) {
+            playbackTimer.cancel();
+            playbackTimer = null;
+        }
+    }
+
+    private void updatePlaybackInfo() {
+        if (player.isPlaying()) {
+            int mCurrentPosition = player.getCurrentPosition();
+            playbackSeekBar.setProgress(mCurrentPosition / 1000);
+            playbackTime.setText(getTimeString(mCurrentPosition));
+        }
+    }
+
+
+    private String getTimeString(long millis) {
+        StringBuffer buf = new StringBuffer();
+
+        int hours = (int)(millis / (1000 * 60 * 60));
+        int minutes = (int)((millis % (1000 * 60 * 60)) / (1000 * 60));
+        int seconds = (int)(((millis % (1000 * 60 * 60)) % (1000 * 60)) / 1000);
+
+        if (hours > 0) {
+            buf.append(String.format("%02d", hours)).append(":");
+        }
+
+        buf.append(String.format("%02d", minutes)).append(":")
+                .append(String.format("%02d", seconds));
```

Comment:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/2aa03aecc11ade0381c1f426b782e030115c4729/app/src/org/commcare/views/widgets/CommCareAudioWidget.java#L198"></span>Implicitly using the default locale is a common source of bugs: Use <code>String.format(Locale, ...)</code> instead</td>
    </tr>
  </tbody>
</table>